compiler/rustc_trait_selection/src/traits/project.rs

```diff
@@ -1856,15 +1874,23 @@ fn confirm_impl_candidate<'cx, 'tcx>(
     let substs =
         translate_substs(selcx.infcx(), param_env, impl_def_id, substs, assoc_ty.defining_node);
     let ty = tcx.type_of(assoc_ty.item.def_id);
+    let term: ty::Term<'tcx> = if INTO_CONST {
```

Comment:
you could check whether the item is a const or type here (`tcx.def_kind(assoc_ty.item.def_id)`) instead of passing it down at the caller site.

Reply:
I think I'd rather pass it down at the caller site for now, it is a very minor optimization, but I feel like that sort of thing can add up, and for a future reader I think it is a bit easier to read.

Reply:
An optimization should have something that is optimized by it. Let's try it without passing it down and if we see a change in perf, then we can document that. Otherwise it makes the code complexer without a performance or API benefit. We don't do this kind of optimization anywhere else.